src/parcsr_mv/par_csr_matop_device.c

```diff
@@ -5,435 +5,644 @@
-   hypre_ParCSRCommPkgSendProcs    (comm_pkg_j) = hypre_ParCSRCommPkgSendProcs(comm_pkg);
-   hypre_ParCSRCommPkgSendMapStarts(comm_pkg_j) = send_jstarts;
-   hypre_ParCSRCommPkgNumRecvs     (comm_pkg_j) = num_recvs;
-   hypre_ParCSRCommPkgRecvProcs    (comm_pkg_j) = hypre_ParCSRCommPkgRecvProcs(comm_pkg);
-   hypre_ParCSRCommPkgRecvVecStarts(comm_pkg_j) = recv_jstarts;
+hypre_CSRMatrix*
+hypre_ConcatDiagAndOffdDevice(hypre_ParCSRMatrix *A)
+{
+   hypre_CSRMatrix *A_diag = hypre_ParCSRMatrixDiag(A);
+   hypre_CSRMatrix *A_offd = hypre_ParCSRMatrixOffd(A);
 
-   /* init communication */
-   /* ja */
-   comm_handle_j = hypre_ParCSRCommHandleCreate_v2(21, comm_pkg_j,
-                                                   HYPRE_MEMORY_DEVICE, d_send_j,
-                                                   HYPRE_MEMORY_DEVICE, d_recv_j);
-   if (want_data)
-   {
-      /* a */
-      comm_handle_a = hypre_ParCSRCommHandleCreate_v2(1, comm_pkg_j,
-                                                      HYPRE_MEMORY_DEVICE, d_send_a,
-                                                      HYPRE_MEMORY_DEVICE, d_recv_a);
-   }
-   else
-   {
-      comm_handle_a = NULL;
-   }
+   hypre_CSRMatrix *B = hypre_CSRMatrixCreate( hypre_CSRMatrixNumRows(A_diag),
+                                               hypre_CSRMatrixNumCols(A_diag) + hypre_CSRMatrixNumCols(A_offd),
+                                               hypre_CSRMatrixNumNonzeros(A_diag) + hypre_CSRMatrixNumNonzeros(A_offd) );
 
-   hypre_TMemcpy(d_recv_i, recv_i, HYPRE_Int, num_rows_recv + 1, HYPRE_MEMORY_DEVICE,
-                 HYPRE_MEMORY_HOST);
+   hypre_CSRMatrixInitialize_v2(B, 0, HYPRE_MEMORY_DEVICE);
 
-   /* create A_ext: on device */
-   A_ext = hypre_CSRMatrixCreate(num_rows_recv, hypre_ParCSRMatrixGlobalNumCols(A), num_nnz_recv);
-   hypre_CSRMatrixI   (A_ext) = d_recv_i;
-   hypre_CSRMatrixBigJ(A_ext) = d_recv_j;
-   hypre_CSRMatrixData(A_ext) = d_recv_a;
-   hypre_CSRMatrixMemoryLocation(A_ext) = HYPRE_MEMORY_DEVICE;
+   hypreDevice_GetRowNnz(hypre_CSRMatrixNumRows(B), NULL, hypre_CSRMatrixI(A_diag),
+                         hypre_CSRMatrixI(A_offd), hypre_CSRMatrixI(B));
 
-   /* output */
-   vrequest = hypre_TAlloc(void *, 3, HYPRE_MEMORY_HOST);
-   vrequest[0] = (void *) comm_handle_j;
-   vrequest[1] = (void *) comm_handle_a;
-   vrequest[2] = (void *) A_ext;
+   hypreDevice_IntegerExclusiveScan(hypre_CSRMatrixNumRows(B) + 1, hypre_CSRMatrixI(B));
 
-   *request_ptr = (void *) vrequest;
+   const dim3 bDim = hypre_GetDefaultDeviceBlockDimension();
+   const dim3 gDim = hypre_GetDefaultDeviceGridDimension(hypre_CSRMatrixNumRows(A_diag), "warp", bDim);
 
-   /* free */
-   hypre_TFree(send_i,     HYPRE_MEMORY_HOST);
-   hypre_TFree(recv_i,     HYPRE_MEMORY_HOST);
-   hypre_TFree(d_send_i,   HYPRE_MEMORY_DEVICE);
-   hypre_TFree(d_send_map, HYPRE_MEMORY_DEVICE);
+   /* WM: Q - need to pass variables intead of hypre_CSRMatrix...() ? Why? */
```

Comment:
I am also curious

Reply:
After further discussion with Brian on this, it seems that this is a truth about sycl. If you have a heap-based pointer to a struct, say a hypre_CSRMatrix *A, that was allocated with a regular malloc, then it's not allowed to pass things like A->data to sycl kernel launches. Note that passing A->size or other things that aren't pointers to allocated data is OK. This is just something I'll have to keep in mind as I continue porting code.